File: Hard/297. Serialize and Deserialize Binary Tree/serialBST.py
"""
Serialization is the process of converting a data structure or object into a sequence of bits so that it can be stored in a file or memory buffer, or transmitted across a network connection link to be reconstructed later in the same or another computer environment.

Design an algorithm to serialize and deserialize a binary tree. There is no restriction on how your serialization/deserialization algorithm should work. You just need to ensure that a binary tree can be serialized to a string and this string can be deserialized to the original tree structure.

Clarification: The input/output format is the same as how LeetCode serializes a binary tree. You do not necessarily need to follow this format, so please be creative and come up with different approaches yourself.

The number of nodes in the tree is in the range [0, 1E4].
-1000 <= Node.val <= 1000

"""
import heapq

# Definition for a binary tree node.
class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Codec:

    def serialize(self, root):
        """Encodes a tree to a single string.
        
        :type root: TreeNode
        :rtype: str
        """
        if not root:
            return []

        treeToString = []
        queue = [root]

        while queue != []:
            temp = queue[0]
            queue.remove(temp)
            treeToString.append(str(temp.val))
            if temp.left:
                queue.append(temp.left)
            else:
                treeToString.append("None")
            if temp.right:
                queue.append(temp.right)
            else:
                treeToString.append("None")

        return ",".join(treeToString)
        

    def deserialize(self, data):
        """Decodes your encoded data to tree.
        
        :type data: str
        :rtype: TreeNode
        """
        data = data.split(",")
        root = TreeNode(int(data[0]))
        queue = [root]
        
            


# Your Codec object will be instantiated and called as such:
# ser = Codec()
# deser = Codec()
# ans = deser.deserialize(ser.serialize(root))

# //////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
# ////////////                       DO NOT INCLUDE BELOW LINES IN LEETCODE SUBMISSION                      ////////////
# //////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
def testCases():
    root = TreeNode(1)
    root.left, root.right = TreeNode(2), TreeNode(3)
    root.right.left = TreeNode(4)
    root.right.right = TreeNode(5)

    output = [1,2,3,None,None,4,5]

    ser = Codec().serialize(root) 
    deser = Codec().deserialize(ser)
        
    
def main():
    testCases()

if __name__ == '__main__':
    main()
        

File: Hard/297. Serialize and Deserialize Binary Tree/serialBST.cpp
#include <vector>
#include <string>
#include <queue>
#include <iostream>
using namespace std;


// Definition for a binary tree node.
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Codec {
public:

    // Encodes a tree to a single string.
    string serialize(TreeNode* root) {
        if (root == NULL) {
            return "";
        }
        string treeToString = to_string(root->val);
        treeToString.append(",");
        queue<TreeNode*> q;
        q.push(root);

        while (!q.empty()) {
            TreeNode* temp = q.front();
            q.pop();
            vector<TreeNode*> children {temp->left, temp->right};

            for (TreeNode* child : children) {
                if (child) {
                    q.push(child);
                    string value = to_string(child->val);
                    value += ",";
                    treeToString += value;
                }
                else {
                    treeToString += "None,";
                }
            }
        }

        return treeToString;
    }

    // Decodes your encoded data to tree.
    TreeNode* deserialize(string data) {
        if (data.length() == 0) return nullptr;
        int pos = 0;
        vector<string> dataToList;
        string token;

        // Convert string delimited by "," to a vector of strings
        // See program 4 here https://www.javatpoint.com/how-to-split-strings-in-cpp
        while ((pos = data.find(",")) < data.size()) {
            token = data.substr(0, pos);
            dataToList.push_back(token);
            data.erase(0, pos + 1);
        }
        
        int n = dataToList.size();
        if (n <= 1) return nullptr;
        TreeNode* root = new TreeNode(stoi(dataToList[0]));
        queue<TreeNode*> q;
        q.push(root);
        int i = 1; int val;
        TreeNode* temp;

        while (i < n && !q.empty()) {
            temp = q.front();
            q.pop();
            if (dataToList[i] != "None") {
                val = stoi(dataToList[i]);
                temp->left = new TreeNode(val);
                q.push(temp->left);
            } 
            i++;

            if (i < n && dataToList[i] != "None") {
                val = stoi(dataToList[i]);
                temp->right = new TreeNode(val);
                q.push(temp->right);
            }
            i++;
        }
        return root;
    }
};

// Your Codec object will be instantiated and called as such:
// Codec ser, deser;
// TreeNode* ans = deser.deserialize(ser.serialize(root));


int main() {
    TreeNode* root = new TreeNode(1);
    root->left = new TreeNode(2);
    root->right = new TreeNode(3);
    root->right->left = new TreeNode(4);
    root->right->right = new TreeNode(5);

    Codec solution;
    string ser = solution.serialize(root);
    TreeNode* deser = solution.deserialize(ser);

    return 0;
}